### PR TITLE
🐛[Story video] Start managing new video source when replaced by mediapool

### PR DESCRIPTION
**Pull request by mszylkowski**
When an `amp-video` is managed by the BitrateManager the child `video` can be replaced by the mediapool, but the new `video` will not contain the waiting listeners that trigger the downgrades.

When a video is replaced by the pool it calls `resetOnDomChange`, which needs to retrigger a `getBitratemanager(this.win).manage(newVideo)` as well as clean up the listeners from the placeholder video.

Test checks for video replacement to be managed

### CODE REVIEW

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 1]
```diff
@@ -241,6 +241,20 @@ export class BitrateManager {
       video.load();
     }
   }
+
+  /**
+   * Called when a video is replaced.
+   * @param {!Element} oldVideo
+   * @param {!Element} newVideo
+   */
+  videoReplaced(oldVideo, newVideo) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:250]: In terms of API I think it would be clearer and easier to maintain/use if we had `manage()` and `unmanage()`

File: extensions/amp-video/0.1/amp-video.js [hunk 2]
```diff
@@ -213,9 +213,7 @@ class AmpVideo extends AMP.BaseElement {
 
     this.video_ = element.ownerDocument.createElement('video');
     // Manage video if the sources contain bitrate or amp-orig-src will be expanded to multiple bitrates.
-    if (
-      this.element.querySelector('source[data-bitrate], source[amp-orig-src]')
-    ) {
+    if (this.hasAnyCachedSources_() || this.hasAnyBitrateSources_()) {
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/amp-video.js:216]: Is this change needed?

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 3]
```diff
@@ -241,6 +241,20 @@ export class BitrateManager {
...
+    if (oldVideo.waitUnlisten_) {
+      oldVideo.waitUnlisten_();
+    }
+    this.videos_ = this.videos_.filter((v) => v.deref() !== oldVideo);
+    this.manage(newVideo);
+    this.sortSources_(newVideo);
```
Comment (gmajoulet) [inline extensions/amp-video/0.1/flexible-bitrate.js:256]: The code that adds sources should call `changedSources` itself, this line would be introducing a different way to use this Class where you don't have to call it manually. Let's keep all code paths/usage similar.

File: extensions/amp-video/0.1/amp-video.js (hunk 2 above)
Comment (mszylkowski) [reply extensions/amp-video/0.1/amp-video.js:216]: It's more accurate since the `hasAnyCahchedSources` method contains more than just a querySelector

**Review by mszylkowski — COMMENTED** [review]
@gmajoulet Added a test to check for this feature

File: extensions/amp-video/0.1/flexible-bitrate.js [hunk 4]
```diff
@@ -91,7 +91,8 @@ export class BitrateManager {
    * @param {!Element} video
    */
   manage(video) {
-    onNontrivialWait(video, () => {
+    video.isBitrateManaged = true;
```
Comment (newmuis) [inline extensions/amp-video/0.1/flexible-bitrate.js:94]: I can't make a comment on a line you didn't change, but does MediaPool call `video.changedSources()`?

Comment (mszylkowski) [reply extensions/amp-video/0.1/flexible-bitrate.js:94]: When mediapool generates the `Sources` object it calls `video.changedSources()`.

Check https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/sources.js#L86-L88

File: extensions/amp-story/1.0/media-pool.js [hunk 5]
```diff
@@ -516,6 +517,11 @@ export class MediaPool {
       new SwapIntoDomTask(placeholderEl)
     ).then(
       () => {
+        if (placeholderEl.isBitrateManaged) {
+          const bitrateManager = getBitrateManager(this.win_);
+          bitrateManager.unmanage(placeholderEl);
```
Comment (newmuis) [inline extensions/amp-story/1.0/media-pool.js:522]: I think in our current architecture, you'd want to create two new MediaTasks (manage and unmanage bitrate management), then enqueue those tasks here.  Having a single task queue for these helps us ensure that the media operations we care about all go through a centralized code path and happen in the order we want them to.

File: extensions/amp-story/1.0/media-pool.js [hunk 6]
```diff
@@ -516,6 +517,11 @@ export class MediaPool {
       new SwapIntoDomTask(placeholderEl)
     ).then(
       () => {
+        if (placeholderEl.isBitrateManaged) {
```
Comment (newmuis) [inline extensions/amp-story/1.0/media-pool.js:520]: I think you're going to need to do the opposite of these when swapping out of dom (unmanage the pool element and manage the placeholder)

**Review by newmuis — COMMENTED** [review]
Nice!  One last comment

File: extensions/amp-story/1.0/media-tasks.js [hunk 7]
```diff
@@ -445,3 +445,55 @@ export class SwapOutOfDomTask extends MediaTask {
...
+
+  /** @override */
+  executeInternal(mediaEl) {
+    this.bitrateManager_.manage(mediaEl);
+    return Promise.resolve();
+  }
+
+  /** @override */
+  requiresSynchronousExecution() {
+    return true;
```
Comment (newmuis) [inline extensions/amp-story/1.0/media-tasks.js:471]: Do these actually require synchronous execution?  What would happen if these were executed asynchronously, but still "immediately" (i.e. in the next frame)

Comment (mszylkowski) [reply extensions/amp-story/1.0/media-tasks.js:471]: No need to require sync

**Review by newmuis — APPROVED** [review]

**Comment by gmajoulet** [comment]
Closing in favor of #32618
